README.md
# signalk-to-nmea2000

[![Greenkeeper badge](https://badges.greenkeeper.io/sbender9/signalk-to-nmea2000.svg)](https://greenkeeper.io/)

Plugin to convert Signal K to NMEA2000

Requires that toChildProcess be set to nmea2000out for the actisense execute provider:

```
     {
      "id": "actisense",
      "pipeElements": [{
        "type": "providers/execute",
        "options": {
          "command": "actisense-serial /dev/ttyUSB0",
          "toChildProcess": "nmea2000out"
        }
      }
```

or you can configure your N2K connection to use canboatjs in the server admin user interface:
![image](https://user-images.githubusercontent.com/1049678/41557237-ac2e2eea-7345-11e8-8719-bbd18ef832cb.png)



Note that if you're using an NGT-1 to transmit AIS, then you need to use their Windows [NMEA Reader](https://www.actisense.com/wp-content/uploads/2017/07/Actisense-NMEA-Reader-v1.517-Setup.exe_.zip) software to add the pgns (129794, 129038, 129041) in the transmitted list. 


CHANGELOG.md
## Change Log

### v2.13.0 (2021/08/17 13:28 +00:00)
- [#54](https://github.com/SignalK/signalk-to-nmea2000/pull/54) feature: add support for navigation data (129283, 129284) (@ardbiesheuvel)
- [#56](https://github.com/SignalK/signalk-to-nmea2000/pull/56) Ais delta guard (@tkurki)
- [#55](https://github.com/SignalK/signalk-to-nmea2000/pull/55) chore: fix failing test (@tkurki)

### v2.12.0 (2020/12/03 21:36 +00:00)
- [#52](https://github.com/SignalK/signalk-to-nmea2000/pull/52) feature: add Sea/Air Temp PGN (130310) (@benwingrove)

### v2.11.0 (2020/10/30 21:34 +00:00)
- [#47](https://github.com/SignalK/signalk-to-nmea2000/pull/47) feature: add support for n2k alerts (126983, 126985) (@jncarter123)

### v2.10.0 (2020/05/26 14:31 +00:00)
- [#50](https://github.com/SignalK/signalk-to-nmea2000/pull/50) feature: source filtering (@tkurki)
- [#49](https://github.com/SignalK/signalk-to-nmea2000/pull/49) fix: plugin name in server ui (@tkurki)

### v2.9.1 (2020/04/28 19:19 +00:00)
- [#48](https://github.com/SignalK/signalk-to-nmea2000/pull/48) fix: crashing with system time enabled (@sbender9)

### v2.9.0 (2020/03/21 13:38 +00:00)
- [#46](https://github.com/SignalK/signalk-to-nmea2000/pull/46) feature: add support for attitude (127257) (@sbender9)

### v2.8.0 (2020/03/10 18:53 +00:00)
- [#40](https://github.com/SignalK/signalk-to-nmea2000/pull/40) Added Atmospheric Pressure using Environmental Parameters PGN 130311. (@htool)
- [#44](https://github.com/SignalK/signalk-to-nmea2000/pull/44) feture: add a setting to configure how to resend (@sbender9)
- [#42](https://github.com/SignalK/signalk-to-nmea2000/pull/42) fix: show an error message for unknown tank types (@sbender9)
- [#41](https://github.com/SignalK/signalk-to-nmea2000/pull/41) feature: add support for conversions to use subscriptions (@sbender9)

### v2.7.0 (2020/03/03 17:21 +00:00)
- [#39](https://github.com/SignalK/signalk-to-nmea2000/pull/39) feature: add option to periodically resend data to the bus (@sbender9)

### v2.6.0 (2020/02/22 19:36 +00:00)
- [#37](https://github.com/SignalK/signalk-to-nmea2000/pull/37) Feature: Extend support for other PGN 130312 temperature sources (@jncarter123)

### v2.5.0 (2019/07/27 18:43 +00:00)
- [#30](https://github.com/SignalK/signalk-to-nmea2000/pull/30) chore: update to latest canboatjs (@sbender9)
- [#29](https://github.com/SignalK/signalk-to-nmea2000/pull/29) feature: support "greyWater"  keys (@sbender9)

### v2.4.1 (2019/06/04 17:32 +00:00)
- [#26](https://github.com/SignalK/signalk-to-nmea2000/pull/26) fix: bad heading values causing ais encoding to fail (@sbender9)

### v2.4.0 (2019/05/29 12:48 +00:00)
- [#25](https://github.com/SignalK/signalk-to-nmea2000/pull/25)  feature: add Speed Through Water PGN (@jtroyer76)

### v2.3.0 (2019/05/02 12:13 +00:00)
- [#24](https://github.com/SignalK/signalk-to-nmea2000/pull/24) feature: add engine parameters rapid update 127488 (@sbender9)

### v2.2.4 (2018/11/06 20:01 +00:00)
- [#23](https://github.com/SignalK/signalk-to-nmea2000/pull/23) fix; update to support changes to the Instance field names in canboat (@sbender9)

### v2.2.3 (2018/09/13 23:30 +00:00)
- [#22](https://github.com/SignalK/signalk-to-nmea2000/pull/22) fix: alternator voltage is getting set to the wrong value (@sbender9)

### v2.2.2 (2018/09/13 16:31 +00:00)
- [#19](https://github.com/SignalK/signalk-to-nmea2000/pull/19) fix: Engine Parameters, Dynamic not getting generated correctly (@sbender9)

### v2.2.1 (2018/08/10 01:18 +00:00)
- [#18](https://github.com/SignalK/signalk-to-nmea2000/pull/18) fix: saving configuration fails when there is no tank data in signalk (@sbender9)

### v2.2.0 (2018/07/14 17:35 +00:00)
- [#12](https://github.com/SignalK/signalk-to-nmea2000/pull/12) [WIP] chore: update to use the latest plugin api (@sbender9)
- [#16](https://github.com/SignalK/signalk-to-nmea2000/pull/16) feature: implement temperature and dynamic engine parameter sending (@mairas)
- [#15](https://github.com/SignalK/signalk-to-nmea2000/pull/15) doc: add reference to canboatjs & admin ui (@tkurki)
- [#13](https://github.com/SignalK/signalk-to-nmea2000/pull/13) doc: note adding pgns to allowed tx list of ntg-1 (@webmasterkai)

### v2.1.0 (2018/02/06 04:03 +00:00)
- [#11](https://github.com/SignalK/signalk-to-nmea2000/pull/11) feature: add conversion for tank level and capacity (@sbender9)

### v2.0.1 (2018/02/05 23:44 +00:00)
- [#10](https://github.com/SignalK/signalk-to-nmea2000/pull/10) fix: battery conversion not working (@sbender9)

### v2.0.0 (2018/01/30 00:42 +00:00)
- [#8](https://github.com/SignalK/signalk-to-nmea2000/pull/8)  refactor: break conversion out into separate files and use canboatjs (@sbender9)
- [#6](https://github.com/SignalK/signalk-to-nmea2000/pull/6) fix: ais PGNs (@sbender9)
- [#5](https://github.com/SignalK/signalk-to-nmea2000/pull/5) Update dependencies to enable Greenkeeper 🌴 (@greenkeeper[bot])
- [#2](https://github.com/SignalK/signalk-to-nmea2000/pull/2) Add 127250 heading (magnetic) conversion (@tkurki)
- [#1](https://github.com/SignalK/signalk-to-nmea2000/pull/1) Formatting & titles to the ui (@tkurki)

conversions/ais.js
const _ = require('lodash')

const static_keys = [
  "name",
  "design.aisShipType",
  "design.draft",
  "design.length",
  "design.beam",
  "sensors.ais.fromCenter",
  "sensors.ais.fromBow" ,
  "design.draft",
  "registrations.imo"
]

const position_keys = [ 'navigation.position' ]

const static_pgn = 129794
const position_pgn = 129038
const aton_pgn = 129041

const navStatusMapping = {
  'not under command': 2,
  'anchored': 1,
  'moored': 5,
  'sailing': 8,
  'motoring': 0,
  'towing < 200m': 3,
  'towing > 200m': 3,
  'pushing': 3,
  'fishing': 7,
  'fishing-hampered': 7,
  'trawling': 7,
  'trawling-shooting': 7,
  'trawling-hauling': 7,
  'not-under-way': 2,
  'aground': 6,
  'restricted manouverability':3,
  'restricted manouverability towing < 200m': 3,
  'restricted manouverability towing > 200m': 3,
  'restricted manouverability underwater operations': 3,
  'constrained by draft': 4,
  'ais-sart': 14,
  'hazardous material high speed': 9,
  'hazardous material wing in ground': 10
}

module.exports = (app, plugin) => {
  return {
    title: `AIS (${static_pgn}, ${position_pgn}, ${aton_pgn})`,
    sourceType: 'onDelta',
    //outputType: 'buffer',
    optionKey: 'AISv2',
    callback: (delta) => {
      var selfContext = 'vessels.' + app.selfId

      if ( delta.context == selfContext || isN2K(delta) ) {
        return null
      }

      if ( delta.context.startsWith('vessels.') ) {
        var hasStatic = hasAnyKeys(delta, static_keys);
        var hasPosition = hasAnyKeys(delta, position_keys)
        
        if ( !hasStatic && !hasPosition ) {
          return null
        }
    
        var vessel = _.get(app.signalk.root, delta.context)
        var mmsi = _.get(vessel, 'mmsi') || findDeltaValue(delta, 'mmsi');
        
        if ( !mmsi ) {
          return null;
        }
        
      var res = []
        if ( hasPosition ) {
          res.push(generatePosition(vessel, mmsi, delta))
        }
        
        if ( hasStatic ) {
          res.push(generateStatic(vessel, mmsi, delta))
        }
        return res
      } else if ( delta.context.startsWith('atons.') ) {
        var vessel = _.get(app.signalk.root, delta.context)
        var mmsi = _.get(vessel, 'mmsi') || findDeltaValue(delta, 'mmsi');

        if ( !mmsi ) {
          return
        }
        
        return [ generateAtoN(vessel, mmsi, delta) ]
      }
    }
  }
}

function generateStatic(vessel, mmsi, delta) {
  var name = _.get(vessel, "name") || findDeltaValue(delta, 'name');
  
  var type = _.get(findDeltaValue(delta, "design.aisShipType"), "id")
  var callsign = findDeltaValue(delta, "communication.callsignVhf")
  var length = _.get(findDeltaValue(delta, 'design.length'), 'overall')
  var beam = findDeltaValue(delta, 'design.beam')
  var fromCenter = findDeltaValue(delta, 'sensors.ais.fromCenter')
  var fromBow = findDeltaValue(delta, 'sensors.ais.fromBow')
  var draft = _.get(findDeltaValue(delta, 'design.draft'), 'maximum')
  var imo = findDeltaValue(delta, 'registrations.imo')
  var dest = findDeltaValue(delta, 'navigation.destination.commonName')
  /*
  type = _.isUndefined(type) ? 0 : type
  callsign = fillASCII(callsign ? callsign : '0', 7)
  name = fillASCII(name ? name : '0', 20)
  length = length ? length * 10 : 0xffff;
  beam = beam ? beam * 10 : 0xffff;
  draft = _.isUndefined(draft) ? 0xffff : draft * 100
  */

  if ( _.isUndefined(imo) ) {
    imo = 0
  } else {
    var parts = imo.split(imo)
    imo = Number(parts[parts.length-1])
  }

  var fromStarboard
  if ( beam && fromCenter ) {
    fromStarboard = (beam / 2 + fromCenter)
  }
  fromBow = fromBow ? fromBow : undefined

  //2017-04-15T14:58:37.625Z,6,129794,43,255,76,05,28,e0,42,0f,0f,ee,8c,00,39,48,41,33,37,39,35,41,54,4c,41,4e,54,49,43,20,50,52,4f,4a,45,43,54,20,49,49,40,4f,8a,07,18,01,8c,00,fe,06,de,44,00,cc,bf,19,e8,03,52,55,20,4c,45,44,20,3e,20,55,53,20,42,41,4c,40,40,40,40,40,04,00,ff

  mmsi = parseInt(mmsi, 10)

  return {
    pgn: static_pgn,
    'Message ID': 5,
    //'Repeat indicator': 0,
    'User ID': mmsi,
    'Callsign': callsign,
    'Name': name,
    'Type of ship': type,
    'Length': length,
    'Beam': beam,
    'Position reference from Starboard': fromStarboard,
    'Position reference from Bow': fromBow,
    'Draft': draft,
    'Destination': dest,
    'AIS version indicator': 0,
    'DTE': 0,
    'AIS Transceiver information': 0
  }

  /*
  var data = Concentrate2()
      .uint8(0x05)
      .uint32(mmsi)
      .uint32(imo)
      .buffer(callsign)
      .buffer(name)
      .uint8(type)
      .uint16(length)
      .uint16(beam)
      .uint16(fromStarboard)
      .uint16(fromBow)
      .buffer(int8buff([0xff, 0xff, 0xff, 0xff, 0xff, 0xff]))
      .uint16(draft)
      .buffer(dest)
      .buffer(int8buff([0x05,0x00,0xff]))
      .result()
  
  return { pgn: static_pgn, buffer:data }
*/
}

function generatePosition(vessel, mmsi, delta) {
  var position = findDeltaValue(delta, 'navigation.position')

  if ( position && position.latitude && position.longitude ) {
    var cog = findDeltaValue(delta, 'navigation.courseOverGroundTrue')
    var sog = findDeltaValue(delta, 'navigation.speedOverGround')
    var heading = findDeltaValue(delta, 'navigation.headingTrue');
    var rot = findDeltaValue(delta, 'navigation.rateOfTurn')
    var status = findDeltaValue(delta, 'navigation.state')

    if ( !_.isUndefined(status) ) {
      status = navStatusMapping[status]
    }
    if ( _.isUndefined(status) ) {
      status = 0
    }

    if ( cog > Math.PI*2 ) {
      cog = undefined
    }

    if ( heading > Math.PI*2 ) {
      heading = undefined
    }
    
    /*
    cog = _.isUndefined(cog) ? 0xffff : (Math.trunc(cog * 10000))
    sog = _.isUndefined(sog) ? 0xffff : (sog*100);
    heading = _.isUndefined(heading) ? 0xffff : (Math.trunc(heading * 10000))
    rot = _.isUndefined(rot) ? 0x7fff : rot

    var latitude = position.latitude * 10000000;
    var longitude = position.longitude * 10000000;
    */

    /*
      2017-04-15T15:06:37.589Z,4,129038,43,255,28,

      01,
      ae,e7,e0,15, mmsi
      36,5c,76,d2, lon
      93,0b,52,17, lat
      94, RAIM/TS
      4d,e9, COG
      39,01, SOG
      7e,05,01,
      ff,ff, heading
      ff,7f, rat
      01,
      00, Nav Status, reserved
      ff reserved
    */

    //console.log(`${mmsi} ${position.longitude} ${position.latitude} ${cog} ${sog} ${heading} ${rot}`)

    return {
      pgn: position_pgn,
      'Message ID': 1,
      //'Repeat Indicator': 0,
      'User ID': mmsi,
      'Longitude': position.longitude,
      'Latitude': position.latitude,
      'Position Accuracy': 0,
      'RAIM': 0,
      'Time Stamp': 0,
      'COG': cog,
      'SOG': sog,
      'AIS Transceiver information': 0,
      'Heading': heading,
      'Rate of Turn': rot,
      'Nav Status': status
    }

    /*
    mmsi = parseInt(mmsi, 10)
    var data = Concentrate2()
        .uint8(0x01)
        .uint32(mmsi)
        .int32(longitude)
        .int32(latitude)
        .uint8(0x94)
        .uint16(cog)
        .uint16(sog)
        .uint8(0x7e)
        .uint8(0x05)
        .uint8(0x01)
        .uint16(heading)
        .int16(rot)
        .uint8(0xff)
        .uint8(0xff)
        .result()
    
    return { pgn: position_pgn, buffer: data }
    */
  } else {
    return null
  }
}

function generateAtoN(vessel, mmsi, delta) {
  var position = findDeltaValue(delta, 'navigation.position')

  if ( position && position.latitude && position.longitude ) {
    var name = _.get(vessel, "name") || findDeltaValue(delta, 'name');
    var type = _.get(findDeltaValue(delta, "atonType"), "id")
    var length = _.get(findDeltaValue(delta, 'design.length'), 'overall')
    var beam = findDeltaValue(delta, 'design.beam')
    var fromCenter = findDeltaValue(delta, 'sensors.ais.fromCenter')
    var fromBow = findDeltaValue(delta, 'sensors.ais.fromBow')
    var latitude = position.latitude * 10000000;
    var longitude = position.longitude * 10000000;

    /*
    type = _.isUndefined(type) ? 0 : type
    name = fillASCII(name ? name : '0', 20)
    length = length ? length * 10 : 0xffff;
    beam = beam ? beam * 10 : 0xffff;
    */

    var fromStarboard
    if ( beam && fromCenter ) {
      fromStarboard = (beam / 2 + fromCenter)
    }
    fromBow = fromBow ? fromBow * 10 : undefined

      /*
  2017-04-15T15:15:08.461Z,4,129041,43,255,49,15,

  77,3c,3a,3b,
  0d,bf,62,d2,
  b3,5e,60,17,
  f5,
  ff,ff,
  ff,ff,
  ff,ff,
  ff,ff, /from True north egde
  4e,
  0e, 
  00,
  01,
  17,01,
  4e,57,
  20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,40
  */
    
    mmsi = parseInt(mmsi, 10)

    return {
      pgn: aton_pgn,
      'Message ID': 0,
      'Repeat Indicator': 0,
      'User ID': mmsi,
      'Longitude': position.longitude,
      'Latitude': position.latitude,
      'Position Accuracy': 0,
      'RAIM': 0,
      'Time Stamp': 0,
      'Length/Diameter': length,
      'Beam/Diameter': beam,
      'Position Reference from Starboard Edge': fromStarboard,
      'Position Reference from True North Facing Edge': fromBow,
      'AtoN Type': type,
      'AtoN Name': name
    }
    /*
    var data = Concentrate2()
        .uint8(0x15)
        .uint32(mmsi)
        .int32(longitude)
        .int32(latitude)
        .uint8(0xf5)
        .uint16(length)
        .uint16(beam)
        .uint16(fromStarboard)
        .uint16(fromBow)
        .tinyInt(type,5)
        .tinyInt(0, 1)
        .tinyInt(0, 1)
        .tinyInt(0, 1)
        .uint8(0x0e)
        .uint8(0x00)
        .uint8(0x01)
        .uint8(0x17)
        .uint8(0x01)
        .buffer(name)
        .uint8(0x40)
        .result()
    return { pgn: aton_pgn, buffer: data }
    */
  } else {
    return null
  }
    
}

function int8buff(array) {
  return new Buffer(new Uint8Array(array).buffer)
}

function hasAnyKeys(delta, keys) {
  if ( delta.updates ) {
    for ( var i = 0; i < delta.updates.length; i++ ) {
      if (Array.isArray(delta.updates[i].values)) {
        for ( var j = 0; j < delta.updates[i].values.length; j++ ) {
          var valuePath = delta.updates[i].values[j].path
          var value = delta.updates[i].values[j].value

          if ( valuePath == '' ) {
            if ( _.intersection(_.keys(value), keys).length > 0 ) {
              return true
            }
          } else if ( keys.includes(valuePath) ) {
            return true
          }
        }
      }
    }
  }
  return false
}

function findDeltaValue(delta, path) {
  if ( delta.updates ) {
    for ( var i = 0; i < delta.updates.length; i++ ) {
      for ( var j = 0; j < delta.updates[i].values.length; j++ ) {
        var valuePath = delta.updates[i].values[j].path
        var value = delta.updates[i].values[j].value
        if ( valuePath == '' && path.indexOf('.') == -1 ) {
          value =  _.get(value, path)
          if ( value ) {
            return value
          }
        } else if ( path == valuePath ) {
          return value
        }
      }
    }
  }
  return undefined
}

function fillASCII(theString, len)
{
  var res = []
  var i
  for ( i = 0; i < len && i < theString.length; i++ )
  {
    res.push(theString.charCodeAt(i))
  }
  for ( ; i < len; i++ )
  {
    res.push(0x40)
  }
  return new Buffer(new Uint8Array(res).buffer);
}

function isN2K(delta) {
  return false
  var res = false
  if ( delta.updates ) {
    delta.updates.forEach(update => {
      var type = _.get(update, 'source.type')
      if ( type && type == 'NMEA2000' ) {
        res = true
      }
    });
  }
  return res
}
